From d88de21b7b131805cff3133af6167a84bf8b77c3 Mon Sep 17 00:00:00 2001
From: Fan Wu <cechkahn@google.com>
Date: Thu, 2 Jan 2025 11:18:59 +0800
Subject: [PATCH 3/4] Check the permission of the callingUid instead of the
 calling package

Bug: 372671447
Test: atest
Flag: EXEMPT bug fix
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:e719575e92c1e6bd0b9088663e6c909bccf2b007)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:062c56f835bec484f97476f42bacd75058c9fb4a)
Merged-In: Ib36c0a3ba482bcddd53c2c09409ea818e6f43cad
Change-Id: Ib36c0a3ba482bcddd53c2c09409ea818e6f43cad
---
 .../settings/applications/AppInfoBase.java    | 26 ++++++++-----------
 1 file changed, 11 insertions(+), 15 deletions(-)

diff --git a/src/com/android/settings/applications/AppInfoBase.java b/src/com/android/settings/applications/AppInfoBase.java
index 9d098000fbc..51e51fc7483 100644
--- a/src/com/android/settings/applications/AppInfoBase.java
+++ b/src/com/android/settings/applications/AppInfoBase.java
@@ -34,6 +34,7 @@ import android.content.pm.PackageManager.NameNotFoundException;
 import android.hardware.usb.IUsbManager;
 import android.os.Bundle;
 import android.os.IBinder;
+import android.os.RemoteException;
 import android.os.ServiceManager;
 import android.os.UserHandle;
 import android.os.UserManager;
@@ -172,24 +173,19 @@ public abstract class AppInfoBase extends SettingsPreferenceFragment
         if (activity == null) {
             return false;
         }
-        String callingPackageName = null;
         try {
-            callingPackageName = ActivityManager.getService()
-                .getLaunchedFromPackage(activity.getActivityToken());
-        } catch (Exception e) {
-            return false;
-        }
-        if (TextUtils.isEmpty(callingPackageName)) {
-            Log.w(TAG, "Not able to get calling package name for permission check");
-            return false;
-        }
-        if (mPm.checkPermission(Manifest.permission.INTERACT_ACROSS_USERS_FULL, callingPackageName)
-                != PackageManager.PERMISSION_GRANTED) {
-            Log.w(TAG, "Package " + callingPackageName + " does not have required permission "
-                    + Manifest.permission.INTERACT_ACROSS_USERS_FULL);
+            int callerUid = ActivityManager.getService().getLaunchedFromUid(
+                    activity.getActivityToken());
+            if (ActivityManager.checkUidPermission(Manifest.permission.INTERACT_ACROSS_USERS_FULL,
+                    callerUid) != PackageManager.PERMISSION_GRANTED) {
+                Log.w(TAG, "Uid " + callerUid + " does not have required permission "
+                        + Manifest.permission.INTERACT_ACROSS_USERS_FULL);
+                return false;
+            }
+            return true;
+        } catch (RemoteException e) {
             return false;
         }
-        return true;
     }
 
     protected void setIntentAndFinish(boolean appChanged) {
-- 
2.40.1

